source/extensions/network/dns_resolver/getaddrinfo/getaddrinfo.cc

```diff
@@ -0,0 +1,217 @@
+      } else if (ai->ai_family == AF_INET6) {
+        sockaddr_in6 address;
+        memset(&address, 0, sizeof(address));
+        address.sin6_family = AF_INET6;
+        address.sin6_port = 0;
+        address.sin6_addr = reinterpret_cast<sockaddr_in6*>(ai->ai_addr)->sin6_addr;
+        v6_results.emplace_back(DnsResponse(std::make_shared<const Address::Ipv6Instance>(address),
+                                            std::chrono::seconds(60)));
+      }
+    }
+    Api::OsSysCallsSingleton::get().freeaddrinfo(addrinfo_result);
+
+    std::list<DnsResponse> final_results;
+    switch (query.dns_lookup_family_) {
+    case DnsLookupFamily::All: {
+      final_results = std::move(v4_results);
+      final_results.splice(final_results.begin(), v6_results);
+      break;
+    }
+    case DnsLookupFamily::V4Only: {
+      final_results = std::move(v4_results);
+      break;
+    }
+    case DnsLookupFamily::V6Only: {
+      final_results = std::move(v6_results);
+      break;
+    }
+    case DnsLookupFamily::V4Preferred: {
+      if (!v4_results.empty()) {
+        final_results = std::move(v4_results);
+      } else {
+        final_results = std::move(v6_results);
+      }
+      break;
+    }
+    case DnsLookupFamily::Auto: {
+      // This is effectively V6Preferred.
+      if (!v6_results.empty()) {
+        final_results = std::move(v6_results);
+      } else {
+        final_results = std::move(v4_results);
+      }
+      break;
+    }
+    }
+
+    ENVOY_LOG(debug, "getaddrinfo resolution complete for host '{}': {}", query.dns_name_,
+              accumulateToString<Network::DnsResponse>(final_results, [](const auto& dns_response) {
+                return dns_response.addrInfo().address_->asString();
+              }));
+
+    return std::make_pair(ResolutionStatus::Success, final_results);
+  }
+
+  // Background thread which wakes up and does resolutions.
+  void resolveThreadRoutine() {
+    ENVOY_LOG(debug, "starting getaddrinfo resolver thread");
+
+    while (true) {
+      PendingQuerySharedPtr next_query;
+      {
+        absl::MutexLock guard(&mutex_);
+        auto condition = [this]() ABSL_EXCLUSIVE_LOCKS_REQUIRED(mutex_) {
+          return shutting_down_ || !pending_queries_.empty();
+        };
+        mutex_.Await(absl::Condition(&condition));
+        if (shutting_down_) {
+          break;
+        }
+
+        next_query = std::move(pending_queries_.front());
+        pending_queries_.pop_front();
+      }
+
+      ENVOY_LOG(debug, "popped pending query [{}]", next_query->dns_name_);
+
+      addrinfo hints;
+      memset(&hints, 0, sizeof(hints));
+      hints.ai_flags = AI_ADDRCONFIG;
+      hints.ai_family = AF_UNSPEC;
+      // If we don't specify a socket type, every address will appear twice, once
+      // for SOCK_STREAM and one for SOCK_DGRAM. Since we do not return the family
+      // anyway, just pick one.
+      hints.ai_socktype = SOCK_STREAM;
+      addrinfo* addrinfo_result;
```

Comment:
Should we put this into a helper class, which destructor will call the `OsSysCallsImpl::freeaddrinfo`

Reply:
Sure.